src/modules/be-water-smart/be-water-smart.component.ts

```diff
@@ -0,0 +1,686 @@
+   */
+  getAlgorithm(algorithmName: string): Algorithm | undefined {
+    var algorithm = this.algorithms.filter(
+      algorithm => algorithm.name === algorithmName,
+    );
+
+    if (algorithm.length === 1) {
+      return algorithm[0];
+    }
+
+    return undefined;
+  }
+
+  /**
+   * calls bws service to retrieve all trained models
+   */
+  extractModels(): void {
+    this.extractData(() => this.bwsService.getModels(), "MLModels", "models");
+  }
+
+  // ---------- Checkbox Functions ----------
+
+  toggleSelectedMeter(item: any, event: Event, selectedMeters: any): void {
+    const isChecked = (event.target as HTMLInputElement).checked;
+
+    if (isChecked) {
+      selectedMeters.push(item);
+    } else {
+      const index = selectedMeters.findIndex(
+        (meter: {id: any}) => meter.id === item.id,
+      );
+      if (index > -1) {
+        selectedMeters.splice(index, 1); // Remove the item if unchecked
+      }
+    }
+  }
+
+  /**
+   * makes only one trained model selectable at a time for forecasting
+   * @param item variable holding the checkbox information
+   */
+  toggleSelectedModel(item: any) {
+    if (this.selectedModel === item) {
+      this.selectedModel = undefined; // Untick the selected item
+    } else {
+      this.selectedModel = item; // Tick the selected item
+    }
+  }
+
+  // ---------- VirtualMeterList Functions ----------
+
+  /**
+   * creates a new VMeter with an @input name and the id-list of the selected physical meters.
+   * If successful, user gets informed and all global variables get set back.
+   * If failed, user gets informed
+   */
+  addVMeter(selectedMeters: any, selectedMeter: string): void {
+    if (!selectedMeter) {
+      alert("No Name for Virtual Meter!");
+      return;
+    }
+
+    this.bwsService
+      .addVirtualMeterWithId(
+        selectedMeter,
+        this.createSubMeterList(selectedMeters),
+      )
+      .subscribe({
+        next: response => {
+          if (response.hasOwnProperty("virtualMeterId")) {
+            this.extractVMeters();
+            this.selectedPhysicalMeters = [];
+            this.selectedVirtualMeters = [];
+            this.newVMeterName = "";
+            this.newSuperMeterName = "";
+          }
+        },
+        error: error => {
+          console.log(error);
+        },
+      });
+  }
+
+  /**
+   * help function for addVMeter()
+   * @returns a list of all ids which are inside the virtual meter
+   */
+  createSubMeterList(selectedMeters: any): object {
+    let id_list: string[] = [];
+
+    selectedMeters.forEach((item: {id: string}) => {
+      id_list.push(item.id);
+    });
+
+    return {submeterIds: id_list};
+  }
+
+  /**
+   * delete a virtual meter
+   * @param id  name of virtual meter, which functions as it's id
+   * @param index index of meter in arr, to hotreload page
+   */
+  deleteVMeterById(id: string, index: number): void {
+    if (this.isDeleting) {
+      return;
+    }
```

Comment:
This feels a bit inverted, you shouldn't call this function if you don't want to delete anything in the first place